test/RefactoredTestSetup.ts

```diff
@@ -0,0 +1,360 @@
+    await this.setUpTestingWorkspace(testReqConfig.projectConfig);
+    if (testReqConfig.projectConfig.projectShape != ProjectShapeOption.NONE) {
+      await utilities.verifyExtensionsAreRunning(utilities.getExtensionsToVerifyActive());
+      const scratchOrgEdition = testReqConfig.scratchOrgEdition || 'developer';
+      this.updateScratchOrgDefWithEdition(scratchOrgEdition);
+      if (process.platform === 'darwin') this.setJavaHomeConfigEntry();
+      // Extra config needed for Apex LSP on GHA
+      if (testReqConfig.isOrgRequired) await this.setUpScratchOrg(scratchOrgEdition);
+      await utilities.reloadAndEnableExtensions(); // This is necesssary in order to update JAVA home path
+    }
+  }
+
+  public async tearDown(): Promise<void> {
+    await this.checkForUncaughtErrors();
+    try {
+      await utilities.deleteScratchOrg(this.scratchOrgAliasName);
+      await this.deleteScratchOrgInfo();
+      this.projectFolderPath = undefined;
+      this.scratchOrgAliasName = undefined;
+      this.scratchOrgId = undefined;
+    } catch (error) {
+      utilities.log(
+        `Deleting scratch org (or info) failed with Error: ${(error as Error).message}`
+      );
+    }
+  }
+
+  public async setUpTestingWorkspace(projectConfig: ProjectConfig) {
+    utilities.log(`${this.testSuiteSuffixName} - Starting setUpTestingWorkspace()...`);
+    let projectName;
+    switch (projectConfig.projectShape) {
+      case ProjectShapeOption.NEW:
+        if (!fs.existsSync(this.tempFolderPath)) {
+          utilities.createFolder(this.tempFolderPath);
+        }
+        await utilities.generateSfProject(this.tempProjectName, this.tempFolderPath); // generate a sf project for 'new'
+        this.projectFolderPath = path.join(this.tempFolderPath, this.tempProjectName);
+        break;
+
+      case ProjectShapeOption.NAMED:
+        if (projectConfig.githubRepoUrl) {
+          // verify if folder matches the github repo url
+          const repoExists = await utilities.gitRepoExists(projectConfig.githubRepoUrl);
+          if (!repoExists) {
+            this.throwError(`Repository does not exist or is inaccessible: ${projectConfig.githubRepoUrl}`);
+          }
+          const repoName = utilities.getRepoNameFromUrl(projectConfig.githubRepoUrl);
+          projectName = repoName;
+          if (projectConfig.folderPath) {
+            const localProjName = utilities.getFolderName(projectConfig.folderPath);
+            if (localProjName != repoName) {
+              this.throwError(`the local project ${projectConfig.githubRepoUrl} does not match the required Github repo ${projectConfig.githubRepoUrl}`);
+            } else {
+              // if it is a match, use the local folder directly. Local dev use only.
+              this.projectFolderPath = projectConfig.folderPath;
+            }
+          } else {
+            // clone the project from Github URL directly
+            this.projectFolderPath = path.join(this.tempFolderPath, repoName!);
```

Comment:
repoName could be null if something fails with `utilities.getRepoNameFromUrl()`

Reply:
it is checked in `gitRepoExists`, which will see if it is a valid github url or not. If it is, then repoName must exist. Does this make sense?